modules/core/04-channel/keeper/upgrade.go

```diff
@@ -86,23 +89,252 @@ func (k Keeper) WriteUpgradeInitChannel(
+// will be returned. If an error occurs in the callback, 0 will be returned but the upgrade sequence will
+// be incremented.
+func (k Keeper) ChanUpgradeTry(
+	ctx sdk.Context,
+	portID string,
+	channelID string,
+	chanCap *capabilitytypes.Capability,
+	counterpartyChannel types.Channel,
+	counterpartyUpgradeSequence uint64,
+	proposedUpgradeChannel types.Channel,
+	timeoutHeight clienttypes.Height,
+	timeoutTimestamp uint64,
+	proofChannel []byte,
+	proofUpgradeTimeout []byte,
+	proofUpgradeSequence []byte,
+	proofHeight clienttypes.Height,
+) (upgradeSequence uint64, previousVersion string, err error) {
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return 0, "", errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if !collections.Contains(channel.State, []types.State{types.OPEN, types.INITUPGRADE}) {
+		return 0, "", errorsmod.Wrapf(types.ErrInvalidChannelState, "expected one of [%s, %s], got %s", types.OPEN, types.INITUPGRADE, channel.State)
+	}
+
+	if !k.scopedKeeper.AuthenticateCapability(ctx, chanCap, host.ChannelCapabilityPath(portID, channelID)) {
+		return 0, "", errorsmod.Wrapf(types.ErrChannelCapabilityNotFound, "caller does not own capability for channel, port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if proposedUpgradeChannel.Counterparty.PortId != channel.Counterparty.PortId ||
+		proposedUpgradeChannel.Counterparty.ChannelId != channel.Counterparty.ChannelId {
+		return 0, "", errorsmod.Wrap(types.ErrInvalidChannel, "counterparty port ID and channel ID cannot be upgraded")
+	}
+
+	if !channel.Ordering.SubsetOf(proposedUpgradeChannel.Ordering) {
+		return 0, "", errorsmod.Wrap(types.ErrInvalidChannelOrdering, "channel ordering must be a subset of the new ordering")
+	}
+
+	if counterpartyChannel.Ordering != proposedUpgradeChannel.Ordering {
+		return 0, "", errorsmod.Wrapf(types.ErrInvalidChannelOrdering, "channel ordering of counterparty channel and proposed channel must be equal")
+	}
+
+	connectionEnd, err := k.getUpgradeTryConnectionEnd(ctx, portID, channelID, channel)
+	if err != nil {
+		return 0, "", err
+	}
+
+	if connectionEnd.GetState() != int32(connectiontypes.OPEN) {
+		return 0, "", errorsmod.Wrapf(
+			connectiontypes.ErrInvalidConnectionState,
+			"connection state is not OPEN (got %s)", connectiontypes.State(connectionEnd.GetState()).String(),
+		)
+	}
+
+	if connectionEnd.GetCounterparty().GetConnectionID() != counterpartyChannel.ConnectionHops[0] {
+		return 0, "", errorsmod.Wrapf(connectiontypes.ErrInvalidConnection, "unexpected counterparty channel connection hops, expected %s but got %s", connectionEnd.GetCounterparty().GetConnectionID(), counterpartyChannel.ConnectionHops[0])
+	}
+
+	if err := k.connectionKeeper.VerifyChannelState(ctx, connectionEnd, proofHeight, proofChannel, proposedUpgradeChannel.Counterparty.PortId,
```

Comment:
So we are verifying 3 things:
- channel
- sequence 
- timeout

From my understanding, there is no particular reason that sequence and timeout are separated within the provable store (they could be under the same upgrade type). The channel is under its' own path for no strong reason, but one benefit is that you can be certain the counterparty has paused packet processing. There could be a construction where all 3 are under the same provable upgrade type:
```
interface Upgrade {
    proposedChannel Channel
    sequence uint64
    timeout Timeout
}
```

We could add a `k.connectionKeeper.VerifyUpgrade` which does all 3, but I don't have a strong preference, just mentioning as food for thought. The main downside of such a function is passing 3 proofs/values at once. The argument list would be very large:
```go
(ctx, connectionEnd, channel.Counterparty.PortId, channel.Counterparty.ChannelId, proofHeight, proofChannel, counterpartyChannel, proofUpgradeTimeout, upgradeTimeout, proofUpgradeSequence, upgradeSequence
```

After seeing that argument list, I think I actually prefer doing the split here